components/compliance-service/reporting/relaxting/suggestions.go

```diff
@@ -502,6 +502,170 @@ func (backend ES2Backend) getControlSuggestions(client *elastic.Client, typePara
+
+	source, err := searchSource.Source()
+	if err != nil {
+		return nil, errors.Wrap(err, "getControlTagsSuggestions unable to get Source")
+	}
+
+	LogQueryPartMin(esIndex, source, "getControlTagsSuggestions query")
+
+	searchResult, err := client.Search().
+		SearchSource(searchSource).
+		Index(esIndex).
+		FilterPath(
+			"took",
+			"hits.total",
+			"hits.hits._id",
+			"hits.hits._score",
+			"hits.hits.inner_hits").
+		Do(context.Background())
+
+	if err != nil {
+		logrus.Error("getControlTagsSuggestions search failed")
+		return nil, err
+	}
+
+	logrus.Debugf("Search query took %d milliseconds\n", searchResult.TookInMillis)
+
+	type ControlSourceTags struct {
+		ID         string                            `json:"id"`
+		StringTags []ESInSpecReportControlStringTags `json:"string_tags"`
+	}
+	type ControlTagsScore struct {
+		Score      float32
+		StringTags []ESInSpecReportControlStringTags
+	}
+
+	// Using this array to aggregate the unsorted suggestions from the response
+	foundControlTags := make([]ControlTagsScore, 0)
+	if searchResult != nil {
+		for _, hit := range searchResult.Hits.Hits {
+			if hit == nil {
+				continue
+			}
+			for _, inner_hit := range hit.InnerHits {
+				if inner_hit == nil {
+					continue
+				}
+				for _, hit2 := range inner_hit.Hits.Hits {
+					if hit2.Source == nil {
+						continue
+					}
+					var c ControlSourceTags
+					err := json.Unmarshal(*hit2.Source, &c)
+					if err == nil && c.ID != "" {
+						foundControlTags = append(foundControlTags, ControlTagsScore{
+							float32(*hit2.Score),
+							c.StringTags,
+						})
+					}
+				}
+			}
+		}
+	}
+	sort.Slice(foundControlTags, func(i, j int) bool {
+		return foundControlTags[i].Score > foundControlTags[j].Score
+	})
+
+	// Map the found tag suggestions to remove duplications
+	scoredTagSuggs := make(map[string]*float32)
+	for _, item := range foundControlTags {
+		matches := make([]string, 0)
+		// If the suggestion is for control tag key, we need to find the one that
+		// was the closest match from all the tag keys the control might have
+		if typeParam == "control_tag_key" {
+			for _, controlTag := range item.StringTags {
+				matches = append(matches, controlTag.Key)
+			}
+		}
+
+		// If the suggestion is for control tag values, we need to find the tag that matches the
+		// tag key and then find the closest match from all the values of that tag
+		if typeParam == "control_tag_value" {
+			for _, controlTag := range item.StringTags {
+				if controlTag.Key == controlTagFilterKey {
+					matches = append(matches, controlTag.Values...)
+				}
+			}
+		}
+
+		// Find the best engram match from the array of texts
+		bestMatch := findBestArrayMatch(text, matches)
+		for _, match := range bestMatch {
+			bestMatchScore := item.Score
+			if scoredTagSuggs[match] == nil || *scoredTagSuggs[match] < item.Score {
+				scoredTagSuggs[match] = &bestMatchScore
+			}
+		}
+	}
+
+	suggs := make([]*reportingapi.Suggestion, 0)
```

Comment:
[nit] We can pre-allocate this

```suggestion
	suggs := make([]*reportingapi.Suggestion, 0, len(scoredTagSuggs))
```

It doesn't matter, it is just a habit I try to keep up.

Reply:
Done, cheers!